### PR TITLE
Add adminClient for topics handling

### PR DESCRIPTION
**Pull request by elakito**
https://cloud-native.slack.com/archives/CMH3Q3SNP/p1588581952045000

Adding adminClient to handle additional rest operations such as GET /topics.

### CODE REVIEW

**Comment by scholzj** [comment]
Thanks for the PR. Could you please do the DCO signoff? As a CNCF project, we need to have it on every commit. Normally, you can fix it just by doing:

```
git commit --amend --signoff
git push --force-with-lease origin work-ac
```

If you click on the _Details_ link next to the DCO check in the staus section of the PR, you will get some more detailed commands. Thanks.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+public abstract class AdminClientEndpoint implements BridgeEndpoint {
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    protected String name;
```
Comment (stanlyDoge) [inline src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:26]: It seems you do not write to the `name`.

Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:26]: I added a line to set this value to "kafka-bridge-admin-client".

**Comment by stanlyDoge** [comment]
Because you've changed OpenAPI definition, you have to rebuild documentation. Please follow the build error message and do 
```
  make docu_api \
    && git add documentation/book/ \
    && git commit -s -m 'Update generated documentation'
```
or similar.

**Comment by elakito** [comment]
I updated this PR with the make docu_api command.

**Comment by elakito** [comment]
still failing at travis-ci but the reason seems to be their file system?
```
java.io.FileNotFoundException: /home/travis/.m2/repository/org/apache/maven/maven-project/2.0.4/maven-project-2.0.4.pom.lastUpdated (No space left on device)
...
[INFO] ------------------------------------------------------------------------
2587[INFO] BUILD FAILURE
2588[INFO] ------------------------------------------------------------------------
2589
```

**Comment by elakito** [comment]
Unrelated to the travis error, but I just noticed that the openapi spec has the wrong response type for GET /topics. I'll fix this.

**Comment by elakito** [comment]
fixed the openapi files and regenerated the doc.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import io.vertx.kafka.admin.TopicDescription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+public abstract class AdminClientEndpoint implements BridgeEndpoint {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:23]: Some Javadoc?

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                );
+                HttpUtils.sendResponse(routingContext, HttpResponseStatus.INTERNAL_SERVER_ERROR.code(),
+                        BridgeContentType.KAFKA_JSON, error.toJson().toBuffer());
+            }
+        });
+    }
+
+    public void doGetTopic(RoutingContext routingContext) {
+        String topicName = routingContext.pathParam("topicname");
+        describeTopics(Arrays.asList(topicName), describeTopicsResult -> {
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:86]: `Collections.singletonList(topicName)`

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                        JsonObject replica = new JsonObject();
+                        replica.put("broker", node.getId());
+                        replica.put("leader", leaderId == node.getId());
+                        replica.put("in_sync", insyncSet.contains(node.getId()));
+                        replicasArray.add(replica);
+                    });
+                    partition.put("replicas", replicasArray);
+                    partitionsArray.add(partition);
+                });
+                root.put("partitions", partitionsArray);
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:113]: I see the OpenAPI spec having a "configs" map with the topic configuration overrides. Where is it filled? You should be able to get this information using the `describeConfigs` method on the Vert.x Kafka admin client.

File: src/main/resources/openapi.json [hunk 5]
```diff
@@ -1564,7 +1625,129 @@
...
+            },
+            "TopicMetadata": {
+                "title": "TopicMetadata",
+                "type": "object",
+                "properties": {
+                    "name": {
+                        "description": "Name of the topic",
+                        "type": "string"
+                    },
+                    "configs": {
```
Comment (ppatierno) [inline src/main/resources/openapi.json:1637]: it seems to be not retrieved from the code.

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        CompletableFuture<Boolean> producer = new CompletableFuture<>();
+        producerService()
+                .sendRecordsRequest(topic, root, BridgeContentType.KAFKA_JSON_JSON)
+                .sendJsonObject(root, ar1 -> {
+                    context.verify(() -> {
+                        assertThat(ar1.succeeded(), is(true));
+                        producer.complete(true);
+                    });
+                });
+        producer.get(TEST_TIMEOUT, TimeUnit.SECONDS);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:47]: why you are producing message?

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:86]: ditto as above

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+                        for (int i = 0; i < 2; i++) {
+                            JsonObject partition = partitions.getJsonObject(i);
+                            assertThat(partition.getInteger("partition"), is(i));
+                            assertThat(partition.getInteger("leader"), is(1));
+                            JsonArray replicas = partition.getJsonArray("replicas");
+                            assertThat(replicas.size(), is(1));
+                            JsonObject replica0 = replicas.getJsonObject(0);
+                            assertThat(replica0.getInteger("broker"), is(1));
+                            assertThat(replica0.getBoolean("leader"), is(true));
+                            assertThat(replica0.getBoolean("in_sync"), is(true));
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:109]: missing "configs" section check which seems to be missing in the code anyway.

**Comment by ppatierno** [comment]
@laidan6000 can you take a look from a documentation point of view in the OpenAPI specification please?

File: documentation/book/api/paths.adoc [hunk 9]
```diff
@@ -936,6 +936,32 @@ Check if the bridge is ready and can accept requests.
 |===
```
Comment (laidan6000) [inline documentation/book/api/paths.adoc:938]: This is great! 
What do you think about the ordering of the methods in the OpenAPI spec? It seems odd to document the /Topics endpoints after the GET methods for `healthy`, `openapi`, and `ready`. I think the following order is more logical:
```
# ...

GET /topics * NEW

GET /topics/{topicname} * NEW

POST /topics/{topicname}

POST /topics/{topicname}/partitions/{partitionid}

GET /healthy

GET /openapi

GET /ready
```

**Comment by laidan6000** [comment]
@elakito - Please can you also update the [Tags](https://strimzi.io/docs/bridge/latest/#tags) section in the OpenAPI specs.

Edit the bullet for Topics:

> * Topics : Topic operations to send messages to a specified topic or topic partition, optionally including message keys in requests. You can also retrieve topics and topic metadata.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Override
+    public BridgeEndpoint closeHandler(Handler<BridgeEndpoint> endpointCloseHandler) {
+        this.closeHandler = endpointCloseHandler;
+        return this;
+    }
+
+    @Override
+    public void open() {
+        // create an admin consumer to handle topic and partition queries
```
[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:53]: ok. I'll change it.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        KafkaConfig kafkaConfig = this.bridgeConfig.getKafkaConfig();
+        Properties props = new Properties();
+        props.putAll(kafkaConfig.getConfig());
+        props.putAll(kafkaConfig.getConsumerConfig().getConfig());
```
[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:57]: I had this consumer config line from the initial code where I used a consumer instance instead. I'll replace it with a dedicated admin config there.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java (hunk 2 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:23]: ok

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java (hunk 3 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:86]: ok

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java (hunk 4 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:113]: ok. I added this call to fill the config.

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java (hunk 8 above)
Comment (elakito) [reply src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:109]: added the configs retrieval and check one of its values.

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java (hunk 6 above)
Comment (elakito) [reply src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:47]: kafkaCluster.createTopic at the beginning of the test adds a topic to Zk and it will take some time until the broker is notified/assigned to it. Sending a message and wait for its delivery as done here makes sure that the topic is known to the broker. Without this message sending and wait part, the test often fails with UnknownTopicOrPartitionException unless Thread.sleep is added to wait for a bit. Please let me know what we should do.

File: documentation/book/api/paths.adoc [hunk 12]
```diff
@@ -936,6 +936,32 @@ Check if the bridge is ready and can accept requests.
...
 
 
+[[_listtopics]]
+=== GET /topics
+
+==== Description
+Retrieves a list of topics.
```
[the remark this answers is not in the document]
Comment (elakito) [reply documentation/book/api/paths.adoc:943]: ok

File: documentation/book/api/paths.adoc [hunk 13]
```diff
@@ -1048,6 +1074,82 @@ __required__||<<_producerrecordlist,ProducerRecordList>>
 ----
 
 
+[[_gettopic]]
+=== GET /topics/{topicname}
+
+==== Description
+Gets metadata about a given topic.
```
[the remark this answers is not in the document]
Comment (elakito) [reply documentation/book/api/paths.adoc:1081]: ok

File: documentation/book/api/paths.adoc [hunk 14]
```diff
@@ -1048,6 +1074,82 @@ __required__||<<_producerrecordlist,ProducerRecordList>>
...
+
+
+==== Parameters
+
+[options="header", cols=".^2a,.^3a,.^9a,.^4a"]
+|===
+|Type|Name|Description|Schema
+|**Path**|**topicname** +
+__required__|Name of the topic to which you want to send records.|string
```
[the remark this answers is not in the document]
Comment (elakito) [reply documentation/book/api/paths.adoc:1090]: ahh. This text is shared by both Post and Get, hence the original text in the context of Posting. I think we can formulate it to apply for both cases. How about "Name of the topic that you want to either retrieve metadata for or send records to."?

**Comment by elakito** [comment]
thank you for all your feedback. I'll push a commit shortly.

**Comment by elakito** [comment]
I added a new commit to resolve the points that you raised. I was not sure if I should overwrite the previous commit or create a new commit to make the updated changes clearly visible. Let me know about additional change requests or whether to merge the commits into one commit.

By the way, I have another branch derived from this branch with the rest of the missing topic related get operations. Operations listPartition and getPartition to get the portion info are ready but the operation to retrieve the offsets summary isn't ready, as the information is not available over adminClient. I think we would need a consumer client. How do you think?

File: documentation/book/api/paths.adoc (hunk 9 above)
Comment (elakito) [reply documentation/book/api/paths.adoc:938]: I think we have no control over the ordering because the spec is currently written in json and the ordering of its entries is not preserved in general. I suppose, one way to preserve the ordering is to write the spec in yaml.

File: src/main/resources/openapi.json (hunk 5 above)
Comment (elakito) [reply src/main/resources/openapi.json:1637]: added the configs

File: documentation/book/api/paths.adoc (hunk 9 above)
Comment (laidan6000) [reply documentation/book/api/paths.adoc:938]: OK, thanks for looking into it. It's not a big deal, so let's leave the order unchanged.

File: documentation/book/api/paths.adoc (hunk 14 above)
[the remark this answers is not in the document]
Comment (laidan6000) [reply documentation/book/api/paths.adoc:1101]: Ah, OK. Your text is great, but please add a comma after "for".

**Comment by elakito** [comment]
I resolved a conflict which recently came into master and force pushed the change. Could you review it?

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                HttpUtils.sendResponse(routingContext, HttpResponseStatus.INTERNAL_SERVER_ERROR.code(),
+                        BridgeContentType.KAFKA_JSON, error.toJson().toBuffer());
+            }
+        });
+    }
+
+    public void doGetTopic(RoutingContext routingContext) {
+        String topicName = routingContext.pathParam("topicname");
+
+        Future<Map<String, TopicDescription>> describeTopicsFuture = Future.future();
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:93]: lately, we moved to use `Promise` instead of `Future`. Can you change here as well, please?

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        this.describeTopics(Collections.singletonList(topicName), describeTopicsFuture);
+        Future<Map<ConfigResource, Config>> describeConfigsFuture = Future.future();
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:95]: ditto as above

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                root.put("name", topicName);
+                List<ConfigEntry> configEntries = configDescriptions.values().iterator().next().getEntries();
+                if (configEntries.size() > 0) {
+                    JsonObject configs = new JsonObject();
+                    configEntries.forEach(configEntry -> {
+                        configs.put(configEntry.getName(), configEntry.getValue());
+                    });
+                    root.put("configs", configs);
+                }
+                TopicDescription description = topicDescriptions.values().iterator().next();
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:114]: as an additional check we can do something like this:

```java
TopicDescription description = topicDescriptions.get(topicName);
```

File: src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java (hunk 6 above)
Comment (ppatierno) [reply src/test/java/io/strimzi/kafka/bridge/http/AdminClientTest.java:49]: Got it, let's stay with the producer.

File: src/test/java/io/strimzi/kafka/bridge/http/HttpCorsTests.java [hunk 18]
```diff
@@ -46,6 +48,7 @@
     static {
         config.put(HttpConfig.HTTP_CONSUMER_TIMEOUT, timeout);
         config.put(BridgeConfig.BRIDGE_ID, "my-bridge");
+        config.put(KafkaConfig.KAFKA_CONFIG_PREFIX + ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/HttpCorsTests.java:51]: why this addition? It doesn't seem to be related to admin stuff.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * Base class for admin client endpoint
+ *
```
[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:27]: ok

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+ */
+public abstract class AdminClientEndpoint implements BridgeEndpoint {
+    protected final Logger log = LoggerFactory.getLogger(getClass());
```
[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:30]: I don't know what I should do here. I am missing something. Could you describe it differently?

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Constructor
+     *
+     * @param vertx Vert.x instance
+     * @param bridgeConfig Bridge configuration
+     */
+    public AdminClientEndpoint(Vertx vertx, BridgeConfig bridgeConfig) {
+        this.vertx = vertx;
+        this.name = "kafka-bridge-admin-client";
```
[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:48]: ok

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java (hunk 15 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:93]: ok

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java (hunk 16 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:95]: ok

File: src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java (hunk 17 above)
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpAdminClientEndpoint.java:114]: ok. I switched to use get(topicName) and added an NPE guard.

File: src/test/java/io/strimzi/kafka/bridge/http/HttpCorsTests.java (hunk 18 above)
Comment (elakito) [reply src/test/java/io/strimzi/kafka/bridge/http/HttpCorsTests.java:51]: It is related, as adminClient is now initialized when HttpBridge is started. Unlike the other HttpBridge tests, HttpCorsTests doesn't use HttpBrdigeTestBase which sets the bootstrap address, so this bootstrap address is set here.

**Comment by elakito** [comment]
@ppatierno thanks for your comments. I force-pushed the requested changes except for one with the below comment

> can we use AdminClientEndpoint.class as we ...

I didn't get what you meant. Could you explain it to me? Thanks.

File: src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (stanlyDoge) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:29]: I think Paolo meant this: `protected final Logger log = LoggerFactory.getLogger(AdminClientEndpoint.class);`

[the remark this answers is not in the document]
Comment (elakito) [reply src/main/java/io/strimzi/kafka/bridge/AdminClientEndpoint.java:29]: Ah. ok, I got it.

**Comment by elakito** [comment]
I changed the logger's usage and force-pushed the change.

**Comment by ppatierno** [comment]
@elakito sorry again, after merging another PR now there is a simple conflict to fix! You should rebase against the new master and fix the conflict then I will approve and merge without any other request for you ;-)

**Comment by ppatierno** [comment]
@elakito I was able to fix the conflicts by myself, I am going to approve and merge the PR.
